xls/third_party/machsuite/fft/fft_reference.cc

```diff
@@ -0,0 +1,165 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <cstdio>
+#include <filesystem>
+#include <iostream>
+
+#include "absl/flags/flag.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/numbers.h"
+#include "xls/common/file/filesystem.h"
+#include "xls/common/file/get_runfile_path.h"
+#include "xls/common/init_xls.h"
+#include "xls/common/integral_types.h"
+#include "xls/common/logging/logging.h"
+#include "xls/common/status/status_macros.h"
+
+#define FFT_SIZE 1024
+
+ABSL_FLAG(bool, hex, false, "Display float data in hex format.");
+
+constexpr const char kRealInPath[] = "xls/examples/fft/test_data/real_in";
+constexpr const char kImgInPath[] = "xls/examples/fft/test_data/img_in";
+constexpr const char kRealTwidInPath[] = "xls/examples/fft/test_data/real_twid_in";
+constexpr const char kImgTwidInPath[] = "xls/examples/fft/test_data/img_twid_in";
+
+namespace xls {
+namespace {
+
+using xls::GetXlsRunfilePath;
+using xls::GetFileContents;
+
+// Note: This reference uses full IEEE-compliant floating-
+// point operations, whereas the XLS version uses floating-
+// point operatoins that do not support subnormal numbers.
+// This version could be modified to flush subnormal numbers
+// to guaruntee equivalence for all inputs. However, this is
+// not necessary for the input tested.
+
+void fft(float real[FFT_SIZE], float img[FFT_SIZE], float real_twid[FFT_SIZE/2], float img_twid[FFT_SIZE/2]){
+    int even, odd, span, log, rootindex;
+    float temp;
+    log = 0;
+
+    for(span=FFT_SIZE>>1; span; span>>=1, log++){
+        for(odd=span; odd<FFT_SIZE; odd++){
+            odd |= span;
+            even = odd ^ span;
+
+            temp = real[even] + real[odd];
+            real[odd] = real[even] - real[odd];
+            real[even] = temp;
+
+            temp = img[even] + img[odd];
+            img[odd] = img[even] - img[odd];
+            img[even] = temp;
+
+            rootindex = (even<<log) & (FFT_SIZE - 1);
+            if(rootindex){
+                temp = real_twid[rootindex] * real[odd] -
+                    img_twid[rootindex]  * img[odd];
+                img[odd] = real_twid[rootindex]*img[odd] +
+                    img_twid[rootindex]*real[odd];
+                real[odd] = temp;
+            }
+        }
+    }
+}
+
+// Load 'data' with the floats contained the file pointed at
+// location 'path'. Expects exactly one float per line.
+absl::Status read_input_data(const char* path, float data[]) {
+  XLS_ASSIGN_OR_RETURN(std::filesystem::path run_path,
+                       GetXlsRunfilePath(path));
+  XLS_ASSIGN_OR_RETURN(std::string data_text, 
+                       GetFileContents(run_path));
+
+  int64 data_idx = 0;
+  for (auto line : absl::StrSplit(data_text, '\n')) {
+    if(line.size()) {
+      XLS_CHECK(absl::SimpleAtof(line, &data[data_idx]));
+      data_idx++;
+    }
+  }
+
+  return absl::OkStatus();
+}
+
+typedef union float_uint_union_t {
```

Comment:
union float_unit_union {
...
};

Reply:
uint as in uint32_t.

Reply:
I meant change the form of the declaration not the name. That is:

union foo {
};

Not:

typedef union foo_t {
} foo;

I'll make the change as it's uploaded internally.